src/client/visualizations/base-visualization/base-visualization.tsx

```diff
@@ -0,0 +1,223 @@
+import * as React from 'react';
+import { $, ply, TimeRange, Expression, Dataset } from 'plywood';
+import {
+  MeasureModeNeeded,
+  Manifest,
+  Measure,
+  VisualizationProps,
+  DatasetLoad,
+  Splits,
+  DataSource,
+  Colors,
+  Resolve,
+  Essence
+} from '../../../common/models/index';
+
+import { SPLIT, SEGMENT, TIME_SEGMENT } from '../../config/constants';
+
+
+export interface BaseVisualizationState {
+  datasetLoad?: DatasetLoad;
+  dragStartTime?: Date;
+  dragTimeRange?: TimeRange;
+  roundDragTimeRange?: TimeRange;
+  dragOnMeasure?: Measure;
+  scrollLeft?: number;
+  scrollTop?: number;
+  hoverTimeRange?: TimeRange;
+  hoverMeasure?: Measure;
+}
+
+export class BaseVisualization<S extends BaseVisualizationState>
+  extends React.Component<VisualizationProps, S> {
+
+  public static id = 'base-visualization';
+  public static title = 'Base Visualization';
+
+  public static measureModeNeed: MeasureModeNeeded = 'single';
+
+  public static handleCircumstance(
+    dataSource: DataSource, splits: Splits, colors: Colors, current: boolean
+  ): Resolve {
+    return Resolve.NEVER;
+  }
+
+  // isMounted was already taken by the superclass
+  protected _isMounted: boolean;
+
+  constructor() {
+    super();
+
+    this.state = this.getDefaultState() as S;
+
+    this.globalMouseMoveListener = this.globalMouseMoveListener.bind(this);
+    this.globalMouseUpListener = this.globalMouseUpListener.bind(this);
+    this.globalKeyDownListener = this.globalKeyDownListener.bind(this);
+  }
+
+  getDefaultState(): BaseVisualizationState {
+    return {
+      datasetLoad: {},
+      dragStartTime: null,
+      dragTimeRange: null,
+      scrollLeft: 0,
+      scrollTop: 0,
+      hoverTimeRange: null,
+      hoverMeasure: null
+    };
+  }
+
+  // Way to get a static property without explicitly specifying the class
+  protected get id(): string {
+    return (this.constructor as any).id;
+  }
+
+  protected makeQuery(essence: Essence): Expression {
+    var { splits, colors, dataSource } = essence;
+    var measures = essence.getEffectiveMeasures();
+
+    var $main = $('main');
+
+    var query = ply()
+      .apply('main', $main.filter(essence.getEffectiveFilter(this.id).toExpression()));
+
+    measures.forEach((measure) => {
+      query = query.performAction(measure.toApplyAction());
+    });
+
+    function makeSubQuery(i: number): Expression {
+      var split = splits.get(i);
+      var splitDimension = dataSource.getDimensionByExpression(split.expression);
+      var { sortAction, limitAction } = split;
+      if (!sortAction) {
+        throw new Error('something went wrong in time series query generation');
+      }
+
+      var segmentName = splitDimension.kind === 'time' ? TIME_SEGMENT : SEGMENT;
```

Comment:
remember how I told you that I hate `TIME_SEGMENT` and it is a giant hack. Well welcome to hating it with me. If you split on time, then click on `barchat` you will get an error. An error that is rooted in this line being a hack